src/main/java/org/broadinstitute/hellbender/utils/runtime/StreamingProcessController.java

```diff
@@ -0,0 +1,598 @@
+    }
+
+    /**
+     * Blocking call to retrieve output from the remote process by prompt synchronization. This call collects
+     * data from the remote process until it receives a prompt on either the stdout or stderr stream, or
+     * the timeout is reached.
+     *
+     * Use isOutputAvailable to ensure the the thread will not block.
+     *
+     * @return ProcessOutput containing a prompt terminated string in either std or stderr
+     * @throws TimeoutException if the timeout is exceeded an no output is available
+     */
+    public ProcessOutput getProcessOutputByPrompt() throws TimeoutException {
+        if (promptForSynchronization == null) {
+            throw new IllegalStateException("A prompt must be specified in order to use prompt-synchronized I/O");
+        }
+        return getOutputSynchronizedBy(promptForSynchronization);
+    }
+
+    /**
+     * Blocking call to retrieve output from the remote process by line. This call collects
+     * data from the remote process until it receives a terminated line on either the stdout or stderr stream, or
+     * the timeout is reached.
+     *
+     * @return ProcessOutput containing a newline terminated string in either std or stderr
+     * @throws TimeoutException if the timeout is exceeded an no output is available
+     */
+    public ProcessOutput getProcessOutputByLine() throws TimeoutException {
+        return getOutputSynchronizedBy("\n");
+    }
+
+    /**
+     * Accumulate output from the target process until either a timeout is reached (no output for timeout duration),
+     * or output containing the {@code #synchronizationString} is detected.
+     *
+     * @param synchronizationString string to synchronize on
+     * @return ProcessOutput containing a {@code synchronizationString} terminated string in either std or stderr
+     * @throws TimeoutException
+     */
+    private ProcessOutput getOutputSynchronizedBy(final String synchronizationString) throws TimeoutException {
+        boolean gotPrompt = false;
+        try (final ByteArrayOutputStream stdOutAccumulator = new ByteArrayOutputStream(CapturedStreamOutput.STREAM_BLOCK_TRANSFER_SIZE);
+             final ByteArrayOutputStream stdErrAccumulator = new ByteArrayOutputStream(CapturedStreamOutput.STREAM_BLOCK_TRANSFER_SIZE))
+        {
+            while (gotPrompt == false) {
+                final ProcessOutput processOutput = getProcessOutput();
+                gotPrompt = scanForSynchronizationPoint(processOutput, synchronizationString);
+                final StreamOutput stdOut = processOutput.getStdout();
+                if (stdOut != null) {
+                    stdOutAccumulator.write(stdOut.getBufferBytes());
+                }
+                final StreamOutput stderr = processOutput.getStderr();
+                if (stderr != null) {
+                    stdErrAccumulator.write(stderr.getBufferBytes());
+                }
+            }
+            stdOutAccumulator.flush();
```

Comment:
I think flushing a ByteArrayOutputStream is a noop.

Reply:
Removed.